graph/patterns/infrastructure-as-code.md

```diff
@@ -0,0 +1,172 @@
+# Infrastructure as code
+
+Microsoft Graph API Design Pattern
+
+*Infrastructure as code (IaC) resource pattern ensures that system resources can be deployed in a reliable, repeatable, and controlled way, normally via idempotent operations.*
+
+## Problem
+
+Infrastructure as code (IaC) defines system resources and topologies in a descriptive manner that allows teams to manage those resources as they would code. 
+Practicing IaC helps teams deploy system resources in a reliable, repeatable, and controlled way. 
+IaC also helps automate deployment and reduces the risk of human error, especially for complex large environments. 
+Customers want to adopt IaC practices for many of the resources managed through Microsoft Graph.
+
+Most resources' creation operations in Microsoft Graph are not idempotent in nature.
+As a consequence, API consumers that want to offer IaC solutions, must create compensation layers that can mimic idempotent behavior. 
+For example, when creating a resource, the compensation layer must check whether the resource first exists, before trying to create or update the resource.
+
+Additionally, IaC code scripts or templates usually employ client-provided names (or keys) to track resources in a predictable manner, whereas [Microsoft Graph guidelines](../GuidelinesGraph.md#behavior-modeling) suggests use of `POST` to create new entities with service-generated keys.
+
+## Solution
+
+The solution is to use an `UPSERT` pattern, to solve for the non-idempotent creation and client-provided naming problems.
+
+* For IaC scenarios, resources must use `UPSERT` semantics with an [alternate key](./alternate-key.md):
+  * Use `PATCH` with a client-provided alternate key.
+  * For a non-existent resource (specified by the alternate key) the service must handle this as a "create". As part of creation, the service must still generate the primary key value.
+  * For an existing resource (specified by the alternate key) the service must handle this as an "update.
+  * Any new alternate key, used for IaC scenarios, be called `uniqueName`.
```

Comment:
I don't think this should be a requirement.  Imagine adding SP today, why would we want them to use uniqueName instead of leveraging the appId?  I think uniqueName is fine as a default in lieu of something that is scenario driven.

Is there value in having another property named uniqueName?

Reply:
It's missing the word "should".  OK how about I qualify this as it should be called `uniqueName` if there isn't already an existing property that could be used as an alternate key.

The reason I want to make the naming recommendation is that it would make things more consistent.  

I had hoped all client provided keys here would have the same name to make the extension simpler - but I guess we'll need some declarative annotations to mark which alternate keys are enabled for upsert, to allow for autogeneration of the bicep types with variably named keys..